Add CLI to help select/run codemods

jfmengels opened the pull request:
I've added a CLI tool to help move from a version of AVA to another.
It kind of looks like this:

![image](https://cloud.githubusercontent.com/assets/3869412/14298459/50cab8d2-fb85-11e5-9b1b-b51d88a1ae3a.png)

I've created a file `codemods.json` which should list the versions that contain codemod-able changes. For now, there is only one entry for v0.14, which is the codemods needed to go from <0.14.0 to 0.14.0. Hopefully, we will later only have create the codemods and modify that file.

We could later have a different question for selecting upgrade codemods (0.13 --> 0.14) or transition codemods (mocha --> ava).

Pretty sure this will not when installed globally (it needs to find `jscodeshift` in the $PATH), would welcome some tips.

jfmengels committed:
Add CLI to help select/run codemods

jamestalmage commented on cli.js:
```diff
@@ -0,0 +1,54 @@
+
+	var result;
+	scripts.forEach(function (script) {
+		result = childProcess.spawnSync('jscodeshift', ['-t', script].concat(files.split(' ')), spawnOptions);
```
Use the full path to `jscodeshift` bin:

``` js
var binPath = require.resolve('jscodeshift/bin/jscodeshift.sh');
```

sindresorhus replied:
That is fragile as they could change the internal name at any time. Better to use https://github.com/sindresorhus/npm-run-path#usage and you can keep spawning it by name. (Like the existing implementation, just in a module and correct.)

jfmengels replied:
:+1:

jamestalmage commented on cli-utils.js:
```diff
@@ -0,0 +1,44 @@
+var semver = require('semver');
```
needs to be added to package.json

jamestalmage commented on cli.js:
```diff
@@ -0,0 +1,54 @@
+function runScripts(scripts, files) {
+	var spawnOptions = {
+		env: assign({}, process.env, {PATH: path.resolve('node_modules/.bin') + ':' + process.env.PATH})
+	};
```
Users should be able to see the stdio output while the command is executing:

add option:

``` js
stdio: 'inherit'
```

Docs are [here](https://nodejs.org/docs/v0.11.13/api/child_process.html#child_process_child_process_spawn_command_args_options). (You need to scroll way down. ridiculous there isn't a closer anchor for such an important option)

jamestalmage commented on package.json:
```diff
@@ -16,14 +17,21 @@
   ],
   "dependencies": {
-    "jscodeshift": "^0.3.19"
+    "inquirer": "^0.12.0",
```
missing `semver`

jfmengels commented:
Actually, do you guys mind if I create a new repo/package for this and we integrate it? I'm thinking this could be of use for quite a few projects.

jamestalmage commented on cli-utils.js:
```diff
@@ -0,0 +1,44 @@
+		return codemod.scripts;
+	});
+
+	return flatten(scripts);
```
You need to resolve the scripts to absolute file paths.

``` js
return flatten(scripts).map(function (script) {
  return path.join(__dirname, script);
});
```

sindresorhus commented:
@jfmengels Can you fix and merge now and do that later? This is the only thing blocking a new AVA release. Definitely a good idea though.

jfmengels commented:
Sure, will do!

sindresorhus commented:
Tests are failing. Maybe also update the docs about running it and what AVA version migrations currently supported.

jamestalmage commented on cli.js:
```diff
@@ -0,0 +1,54 @@
+
+	var scripts = utils.selectScripts(codemods, answers.currentVersion, answers.nextVersion);
+
+	runScripts(scripts, answers.files);
```
let's allow them to pass in globs

``` js
runScripts(scripts, globby.sync(answers.files.split(/\s+/));
```

jfmengels commented:
Should we recommend to install it globally? Makes it easier to run the command.

sindresorhus commented:
@jfmengels Yup, use https://github.com/sindresorhus/meow to handle the input and help stuff.

jamestalmage commented:
Yes. This should definitely be installed globally.

jamestalmage commented:
> meow

He's doing everything with prompts now. So no need for meow yet

jfmengels committed:
Update README.md

sindresorhus commented:
@jamestalmage I know, but `meow` handles other things too, like `--version`, `--help`: https://github.com/sindresorhus/meow#features We should definitely have `--help`.

jfmengels commented:
I've updated the PR. Thanks for all the tips @jamestalmage.

Have updated the readme a bit, can be built on later.

@sindresorhus Fair point for `--version`/`--help`. I'm adding it now (also fixing a bug now that I've added globby).

The way things work now, all files in the project will be listed in the command when launching `jscodeshift`. Will that not break a console limit?

sindresorhus commented:
Note: https://github.com/jamestalmage/ava-codemods/pull/9#discussion_r58757388

sindresorhus commented on cli.js:
```diff
@@ -0,0 +1,56 @@
+
+	var result;
+	scripts.forEach(function (script) {
+		result = childProcess.spawnSync(binPath, ['-t', script].concat(files.split(' ')), spawnOptions);
```
Why are we even spawning it? https://github.com/facebook/jscodeshift/blob/ab22e68ec65a376e01759e75643efb94605fb6e8/bin/jscodeshift.sh#L78-L82

jfmengels replied:
It didn't have a documented API to run it, so I went for the spawn way. The runner is not an exported module, but an executable, so not sure how we could use that one.

sindresorhus replied:
Yeah, we can defer that to later, but would be cleaner to use it I think.

It would be:

``` js
var Runner = require('jscodeshift/dist/Runner');
```

Btw, you could also require the CLI instead of spawning it.

jfmengels replied:
Yeah but you can only do that once right? Since `jscodeshift` only accepts one transformer file at a time (afaik), we need to spawn it multiples times.

But I agree that it would be nice to avoid spawning it...

sindresorhus replied:
Anyways. Doesn't matter right now :)

sindresorhus commented:
We should warn people the following either in the readme or the CLI. Something like:

> Ensure you have a backup of your tests or commit the latest changes before running this.

jfmengels committed:
Add meow, warning messages and simplify spawn path

sindresorhus commented on cli.js:
```diff
@@ -1,30 +1,41 @@
+		Usage
+			$ ava-codemods
+
+		Ensure you have a backup of your tests or commit the latest changes before running this.
```
I don't think it should be here as most won't read the `--help`, and we already have it elsewhere.

jfmengels replied:
I've added it to all three places. Do you want me to remove it here?

sindresorhus replied:
Yeah

sindresorhus commented on cli.js:
```diff
@@ -1,30 +1,41 @@
+		Ensure you have a backup of your tests or commit the latest changes before running this.
+
+		Upgrades available:
+			- 0.13.x to 0.14.x
```
```
- 0.13.x → 0.14.x
```

sindresorhus commented on cli.js:
```diff
@@ -1,30 +1,41 @@
+
+		Ensure you have a backup of your tests or commit the latest changes before running this.
+
+		Upgrades available:
```
`Upgrades available:` => `Available upgrades`

sindresorhus commented on cli.js:
```diff
@@ -1,30 +1,41 @@
 
 function runScripts(scripts, files) {
-	var binPath = require.resolve('jscodeshift/bin/jscodeshift.sh');
+	// var binPath = require.resolve('jscodeshift/bin/jscodeshift.sh');
```
Leftover

jfmengels committed:
Update --help message

jfmengels commented:
All comments taken into account.

jamestalmage commented:
Nice!

jfmengels commented:
:dancer:

sindresorhus commented:
Wondrous‼️ ✨
